core/arch/arm/kernel/thread_a32.S

```diff
@@ -668,48 +566,208 @@ UNWIND(	.cantunwind)
 UNWIND(	.fnend)
 END_FUNC thread_unwind_user_mode
 
-LOCAL_FUNC thread_abort_handler , :
-thread_und_handler:
-UNWIND(	.fnstart)
-UNWIND(	.cantunwind)
+	.macro maybe_restore_mapping
+		/*
+		 * This macro is a bit hard to read due to all the ifdefs,
+		 * we're testing for two different configs which makes four
+		 * differnt combinations.
+		 *
+		 * - With LPAE, and then some extra code if with
+		 *   CFG_CORE_UNMAP_CORE_AT_EL0
+		 * - Without LPAE, and then some extra code if with
+		 *   CFG_CORE_UNMAP_CORE_AT_EL0
+		 */
+
+		/*
+		 * At this point we can't rely on any memory being writable
+		 * yet, so we're using TPIDRPRW to store r0, and if with
+		 * LPAE TPIDRURO to store r1 too.
+		 */
+		write_tpidrprw r0
+
+#ifdef CFG_WITH_LPAE
+		write_tpidruro r1
+
+		read_ttbr0_64bit r0, r1
+		tst	r1, #BIT(TTBR_ASID_SHIFT - 32)
+		beq	11f
+
+#ifdef CFG_CORE_UNMAP_CORE_AT_EL0
+		/* Update the mapping to use the full kernel mode mapping. */
+		sub	r0, r0, #CORE_MMU_L1_TBL_OFFSET
+#endif
```

Comment:
Are we sure there is no offset MSB part to substract from r1 ?
I would think there is none, but it is possible the L1 is above the 4 first GByte ?

Reply:
Thanks, I'll change this into 64-bit arithmetic:
```
                subs    r0, r0, #CORE_MMU_L1_TBL_OFFSET
                sbc     r1, r1, #0
```